### DISCUSSION TITLE
HTTPS with additional authentification

### DISCUSSION THREAD

**Discussion by raddirad** [ipxe/ipxe#689] [General]
Hi

I am playing around with iPXE and wanted to test a boot from https
Here are the configs I changed
```
#define	DOWNLOAD_PROTO_HTTPS	/* Secure Hypertext Transfer Protocol */
....
#define CERT_CMD		/* Certificate management commands */
```

Here's the menu.ipxe
```
#!ipxe
console
#colour --rgb 0xffffff 7
#colour --rgb 0x000000 0
cpair --foreground 7 --background 0 1 
menu Make your choice
item localboot Boot Next Device
item bootimage-tftp Boot Bootimage from TFTP
item bootimage-http Boot Bootimage from HTTP
item bootimage-https-auth VMEX16206
choose os && goto ${os}

:localboot
exit

:bootimage-tftp
kernel ../install-x64 initrd=miniroot-x64 video=vesa:ywrap,mtrr vga=791 quiet splash --no-log console=tty1 console=ttyS0 service=https://192.168.1.1:4447/rpc
initrd ../miniroot-x64 
boot

:bootimage-http
kernel http://192.168.1.1/bootimage/install-x64 initrd=miniroot-x64 video=vesa:ywrap,mtrr vga=791 quiet splash --no-log console=tty1 console=ttyS0 service=https://192.168.1.1:4447/rpc
initrd http://192.168.1.1/bootimage/miniroot-x64 
boot

:bootimage-https-auth
kernel https://myUser:myPassword@192.168.1.1:4447/boot/linux/install-x64 initrd=miniroot-x64 video=vesa:ywrap,mtrr vga=791 quiet splash --no-log console=tty1 console=ttyS0 service=https://192.168.1.1:4447/rpc
initrd https://myUser:myPassword@192.168.1.14:4447/boot/linux/miniroot-x64 
boot
```

The server cert has been compiled in
```
make bin-x86_64-efi/ipxe.efi TRUST=../../ca-cert.cer
```

So when I select the HTTP entry the files are transfered fine. HTTP is implemented with Apache2
However selecting the HTTPS entry with username and password, which works in my browsers, I immediately get a 
```
https://192.168.1.1:4447/boot/linux/install-x64... No such file or directory (https://ipxe.org/2d0c618e)
```
HTTPS access is implemented with WebDAV

Is an additional authentification supported in HTTPS in iPXE?

Regards

**Comment by stappersg** [comment]
On Fri, Jul 22, 2022 at 06:44:34AM -0700, raddirad wrote:
> https://192.168.1.1:4447/boot/linux/install-x64... No such file or directory (https://ipxe.org/2d0c618e)

Text from https://ipxe.org/err/2d0c61

Error: HTTP 404 Not Found

(Error code 2d0c61)
Possible sources

This error originated from one of the following locations within the
iPXE source code:

    net/tcp/httpcore.c (line 1181)

    General advice

    * Try using the latest version of iPXE. Your problem may have
      already been fixed.
    * Try building iPXE with the debug option DEBUG=httpcore



Things to try:

* Check the URL that you are attempting to boot from.

* Try downloading the URL using a web browser, or command-line
  client such as wget.

* Check your certificate trust settings.

**Reply by raddirad** [reply]
With ```DEBUG=httpcore``` I get the addition output
```
HTTP 0xbe518a48 Status 404 not found.
```

The files are accessible via browsers with the given login.

ipxe.efi has been compiled with TRUST=PathToCert/ca-cert.crt
How can I check the certificate trust settings?

**Reply by raddirad** [reply]
So running
```
openssl s_client -connect 192.168.1.1:4447 -debug -CAfile ./ca-cert.cer
```

returned
```
...
SSL handshake has read 2532 bytes and written 399 bytes
Verification: OK
...
```

Should be fine, right?

**Comment by NiKiZe** [comment]
Did you read that error page fully? 😉

Does the iPXE client has full access to the internet? Or is it perhaps limited in some way?

**Reply by raddirad** [reply]
The Client has full internet access

you mean this note on the error page?
```
Note that the same error can occur if the automatic TLS certificate validator does not find a cross-signing certificate on http://ca.ipxe.org/ for the TLS certificate of the server you're trying to connect to.
```

So we should then compile our DER file with ```CERT=ca.crt``` option?

**Reply by raddirad** [reply]
So after adding ```CERT=ca-cert.cer``` to the make command I now get 401 unauthorized but the download still finishes and start the bootimage

**Reply by raddirad** [reply]
removing ```DEBUG=httpcore``` doesn't show any 401 errors, as they are the dafault way of trying to download the desired file. Wget shows the same behaviour.

Can be closed

**Comment by stappersg** [comment]
> Can be closed

With which conclusion?

**Reply by raddirad** [reply]
it works properly, if done correctly. It was more of an understanding problem rather than a technical one.

**Reply by stappersg** [reply]
Please take some time to write down your _correctly done_.

Karma bonus points for pointing to where at the _iPXE documentation_ you would to expect to find it.

In other words: _Help with making it possible for others, like your future self, to reproduce "HTTPS with additional authentification"._

**Reply by raddirad** [reply]
To me the main problem was the understanding what to embed into the ipxe.efi.
I thought that adding the trusted cert was enough. Building it with a pem format cert was compiling fine but then resultet in 404 errors.
Adding the same pem cert file as a CERT parameter solved the issue for me. This was not so clear to me. After reading through the error pages I saw that only pem certs from the iPXE Root CA are trusted.

**Reply by stappersg** [reply]
Only `CERT=...`?
Or is the `TRUST=...` also needed?

**Reply by raddirad** [reply]
I haven't tried without ```TRUST?...``` because documentation says
```
Embedded certificates are not automatically trusted; you will need to specify any [trusted root certificates](https://ipxe.org/crypto#trusted_root_certificates) explicitly using the TRUST=... build parameter. For example:
```

**Reply by stappersg** [reply]
And documented at https://ipxe.org/crypto#embedded_certificates as
```text
make bin/ipxe.iso CERT=ca.crt TRUST=ca.crt
```

(github doesn't allow me to close this)
